add defrag command

kellyselden opened the pull request.

NullVoxPopuli commented on bin/commands/defrag.js:
```diff
@@ -0,0 +1,30 @@
+module.exports = {
+  command: 'defrag',
+  describe: 'synchronize all dependency version discrepancies',
+  builder: {
```
this is very exciting :D 

these options provide much more utility than my one-off script does

NullVoxPopuli commented on src/build-dep-graph.js:
```diff
@@ -106,7 +107,9 @@ async function buildDepGraph({
   await firstPass(workspaceMeta, workspacePackageJson, packageDirs);
-  secondPass(workspaceMeta);
+  if (shouldPruneDeps) {
+    secondPass(workspaceMeta);
```
what does `secondPass` mean? can this be something more meaningful?

NullVoxPopuli commented on src/build-dep-graph.js:
```diff
@@ -75,6 +75,7 @@ function secondPass(workspaceMeta) {
 
 async function buildDepGraph({
   workspaceCwd,
+  shouldPruneDeps = true,
```
what does it mean to prune deps? based on the added test (which sets the option to false) , I'm not sure

kellyselden replied:
before this new feature, monorepo-next was only concerned with packages in the monorepo, and excluded everything else. This brings in the ability to deal with non-monorepo packages for defragging.

NullVoxPopuli commented on src/fs.js:
```diff
@@ -0,0 +1,39 @@
+      newJson = oldJson;
+    }
+
+    let newContents = JSON.stringify(newJson, null, 2) + EOL;
```
:raised_hands:

NullVoxPopuli commented on src/defrag.js:
```diff
@@ -0,0 +1,159 @@
+
+const defaultDependenciesFilter = [];
+
+function firstPass(workspaceMeta, {
```
maybe: `gatherPackageRanges`?

NullVoxPopuli commented on src/defrag.js:
```diff
@@ -0,0 +1,159 @@
+  return allRanges;
+}
+
+function secondPass(allRanges, {
```
maybe: `determineUpdatedPackageRanges`?

NullVoxPopuli commented on src/defrag.js:
```diff
@@ -0,0 +1,159 @@
+  return rangeUpdates;
+}
+
+function thirdPass(workspaceMeta, rangeUpdates, {
```
maybe: `updatePackageRanges`?

NullVoxPopuli commented on src/defrag.js:
```diff
@@ -0,0 +1,159 @@
+  }
+}
+
+async function fourthPass(workspaceMeta) {
```
maybe: `writePackageUpdates`?

NullVoxPopuli commented:
All the tests use the `^` range. for folks who maybe want to pin their deps, how would that work?

kellyselden commented:
> All the tests use the ^ range. for folks who maybe want to pin their deps, how would that work?

I changed an option to `outOfRange` to deal with pinned deps better and added tests. I think the command you now want is `next defrag --out-of-range minor`.

kellyselden commented:
I updated the function names to be more descriptive. Thanks for the callout.
